crates/storage/nippy-jar/src/lib.rs

```diff
@@ -0,0 +1,407 @@
+    /// Compressor if required
+    compressor: Option<Compressors>,
+    #[serde(skip)]
+    /// Data path for file. Index file will be `{path}.idx`
+    path: Option<PathBuf>,
+    /// soon
+    bloom_filter: bool,
+    /// soon
+    phf: bool,
+}
+
+impl NippyJar {
+    /// Creates new [`NippyJar`].
+    pub fn new(columns: usize, bloom_filter: bool, phf: bool, path: &Path) -> Self {
+        NippyJar {
+            version: NIPPY_JAR_VERSION,
+            columns,
+            bloom_filter,
+            phf,
+            compressor: None,
+            path: Some(path.to_path_buf()),
+        }
+    }
+
+    /// Adds [`compression::Zstd`] compression.
+    pub fn with_zstd(mut self, use_dict: bool, max_dict_size: usize) -> Self {
+        self.compressor =
+            Some(Compressors::Zstd(compression::Zstd::new(use_dict, max_dict_size, self.columns)));
+        self
+    }
+
+    /// Loads the file configuration and returns [`Self`].
+    pub fn load(path: &Path) -> Result<Self, NippyJarError> {
+        let mut file = File::open(path)?;
+        let mut obj: Self = bincode::deserialize_from(&mut file)?;
+
+        obj.path = Some(path.to_path_buf());
+        Ok(obj)
+    }
+
+    /// Returns the path from the data file
+    pub fn data_path(&self) -> PathBuf {
+        self.path.clone().expect("exists")
+    }
+
+    /// Returns the path from the index file
+    pub fn index_path(&self) -> PathBuf {
+        let data_path = self.data_path();
+        data_path
+            .parent()
+            .expect("exists")
+            .join(format!("{}.idx", data_path.file_name().expect("exists").to_string_lossy()))
+    }
+
+    /// If required, prepares any compression algorithm to an early pass of the data.
+    pub fn prepare(
+        &mut self,
+        columns: Vec<impl IntoIterator<Item = Vec<u8>>>,
+    ) -> Result<(), NippyJarError> {
+        // Makes any necessary preparations for the compressors
+        if let Some(compression) = &mut self.compressor {
+            compression.prepare_compression(columns)?;
+        }
+        Ok(())
+    }
+
+    /// Writes all necessary configuration to file.
+    pub fn freeze(
+        &mut self,
+        columns: Vec<impl IntoIterator<Item = Vec<u8>>>,
+        total_rows: u64,
+    ) -> Result<(), NippyJarError> {
+        let mut file = self.freeze_check(&columns)?;
+
+        self.freeze_config(&mut file)?;
+
+        // Special case for zstd that might use custom dictionaries/compressors per column
+        // If any other compression algorithm is added and uses a similar flow, then revisit
+        // implementation
+        let mut maybe_zstd_compressors = None;
+        if let Some(Compressors::Zstd(zstd)) = &self.compressor {
+            maybe_zstd_compressors = zstd.generate_compressors()?;
+        }
+
+        // Temporary buffer to avoid multiple reallocations if compressing to a buffer (eg. zstd w/
+        // dict)
+        let mut tmp_buf = Vec::with_capacity(100);
```

Comment:
I think this tmp buf capacity needs to be per data type, otherwise it can be e.g. too much for headers and not enough for transactions?

Reply:
i don't give it much thought (maybe even reduce it). there will be a couple reallocations in a list of thousands items anyway